crates/bevy_ecs/src/storage/table/mod.rs

```diff
@@ -0,0 +1,860 @@
+        }
+        core::mem::forget(_guard); // The allocation was successful, so we don't drop the guard.
+    }
+
+    /// Allocates space for a new entity
+    ///
+    /// # Safety
+    /// the allocated row must be written to immediately with valid values in each column
+    pub(crate) unsafe fn allocate(&mut self, entity: Entity) -> TableRow {
+        self.reserve(1);
+        let len = self.entity_count();
+        self.entities.push(entity);
+        for col in self.columns.values_mut() {
+            col.added_ticks
+                .initialize_unchecked(len, UnsafeCell::new(Tick::new(0)));
+            col.changed_ticks
+                .initialize_unchecked(len, UnsafeCell::new(Tick::new(0)));
+            #[cfg(feature = "track_change_detection")]
+            col.changed_by
+                .initialize_unchecked(len, UnsafeCell::new(Location::caller()));
+        }
+        TableRow::from_usize(len)
+    }
+
+    /// Gets the number of entities currently being stored in the table.
+    #[inline]
+    pub fn entity_count(&self) -> usize {
+        self.entities.len()
+    }
+
+    /// Get the drop function for some component that is stored in this table.
+    #[inline]
+    pub fn get_drop_for(&self, component_id: ComponentId) -> Option<unsafe fn(OwningPtr<'_>)> {
+        self.get_column(component_id)?.data.drop
+    }
+
+    /// Gets the number of components being stored in the table.
+    #[inline]
+    pub fn component_count(&self) -> usize {
+        self.columns.len()
+    }
+
+    /// Gets the maximum number of entities the table can currently store
+    /// without reallocating the underlying memory.
+    #[inline]
+    pub fn entity_capacity(&self) -> usize {
+        self.entities.capacity()
+    }
+
+    /// Checks if the [`Table`] is empty or not.
+    ///
+    /// Returns `true` if the table contains no entities, `false` otherwise.
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.entities.is_empty()
+    }
+
+    /// Call [`Tick::check_tick`] on all of the ticks in the [`Table`]
+    pub(crate) fn check_change_ticks(&mut self, change_tick: Tick) {
+        let len = self.entity_count();
+        for col in self.columns.values_mut() {
+            // SAFETY: `len` is the actual length of the column
+            unsafe { col.check_change_ticks(len, change_tick) };
+        }
+    }
+
+    /// Iterates over the [`ThinColumn`]s of the [`Table`].
+    pub fn iter_columns(&self) -> impl Iterator<Item = &ThinColumn> {
+        self.columns.values()
+    }
+
+    /// Clears all of the stored components in the [`Table`].
+    pub(crate) fn clear(&mut self) {
+        let len = self.entity_count();
+        for column in self.columns.values_mut() {
+            // SAFETY: we defer `self.entities.clear()` until after clearing the columns,
+            // so `self.len()` should match the columns' len
+            unsafe { column.clear(len) };
+        }
+        self.entities.clear();
```

Comment:
I feel like this still double drops. Specifically if you have two columns A and B after successfully calling clear on A if you then panic on B you will leak anything in B that was after but then double drop A.

It feels like rather than trying to make all of the column stuff drop safe simply calling `self.entities.clear()` first and leaking everything in the case of a panic would be safer and simpler.

Reply:
Panics don't necessarily mean the process exits, so that could result in memory leaks.

Reply:
> Panics don't necessarily mean the process exits, so that could result in memory leaks.

You generally have two choices in these situations:

- leak the remaining elements and let the panic continue;

- continue dropping the other elements during the panic
  - this is was the stdlib does e.g. for `Vec::clear` or `Vec::drop`;
  - note however that if other `drop` panic this will result in an abort.

Note though that the second options is significantly harder to (correctly) implement. In particular you'll have to do this for both `ThinColumn` (which currently does leak the remaining elements by setting the `drop` fn to `None` during `clear_element`) and `Table`.

Reply:
To be clear though unless I am mistaken this implementation isn't panic safe due to potentially dropping records in an already cleared column.

So either the "set drop to None to avoid running it and leaking" fix is needed to extend across multiple columns or some other trick is needed to ensure it doesn't call `drop` on a column that was already `clear`ed earlier.

Reply:
Yes, the implementation isn't panic safe, good catch.
I decided to go with just leaking the elements, mostly because Drop impls panicking are so unusual, and the fix is pretty complicated so better left for a follow-up PR.